### PR TITLE
Remove verticesPerElement

### PR DESCRIPTION
**Pull request by jamesmporter**
Addresses #1425.

**To do:**
- [x] Squash and rename commits

### CODE REVIEW

File: packages/d3fc-webgl/src/buffers/baseAttributeBuilder.js [hunk 1]
```diff
@@ -7,13 +7,13 @@ export default () => {
     let normalized = false;
     let stride = 0;
     let offset = 0;
-    let validSize = 0;
+    let hasDataBeenRendered = false;
```
Comment (OliverForeman) [inline packages/d3fc-webgl/src/buffers/baseAttributeBuilder.js]: I'm not sure this name is completely accurate, maybe `hasPropertyChanged` or `hasAttributeChanged` could be more descriptive?

File: packages/d3fc-webgl/src/buffers/elementConstantAttributeBuilder.js [hunk 2]
```diff
@@ -26,10 +26,10 @@ export default () => {
         return projectedData;
     };
 
-    const build = (gl, program, name, verticesPerElement, count) => {
-        base(gl, program, name, verticesPerElement);
+    const build = (gl, program, name, count) => {
+        base(gl, program, name);
 
-        if (base.validSize() >= count) {
+        if (base.hasDataBeenRendered() === true) {
```
Comment (OliverForeman) [inline packages/d3fc-webgl/src/buffers/elementConstantAttributeBuilder.js]:
```suggestion
        if (base.hasDataBeenRendered()) {
```

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js [hunk 3]
```diff
@@ -31,30 +27,33 @@ export default () => {
...
+
         return projectedData;
     };
 
-    const build = (gl, program, name, verticesPerElement, count) => {
+    const build = (gl, program, name, count) => {
         base(gl, program, name);
 
-        if (base.validSize() >= verticesPerElement) {
+        if (base.hasDataBeenRendered() === true) {
```
Comment (OliverForeman) [inline packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]:
```suggestion
        if (base.hasDataBeenRendered()) {
```

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js [hunk 4]
```diff
@@ -31,30 +27,33 @@ export default () => {
...
             return;
         }
 
-        const projectedData = project(verticesPerElement);
+        console.log(data);
```
Comment (OliverForeman) [inline packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]: Console log

**Review by OliverForeman — CHANGES_REQUESTED** [review]

File: packages/d3fc-webgl/src/buffers/baseAttributeBuilder.js (hunk 1 above)
Comment (chrisprice) [reply packages/d3fc-webgl/src/buffers/baseAttributeBuilder.js]: `dirty`?

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js [hunk 5]
```diff
@@ -8,19 +8,15 @@ export default () => {
 
     let data = null;
 
-    const project = verticesPerElement => {
+    const project = () => {
         const components = base.size();
         const offset = base.offset();
-        const componentsPerElement = verticesPerElement * components;
+        const uniqueVerticesPerElement = data.length;
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]: Maybe just remove this?

File: packages/d3fc-webgl/src/buffers/elementConstantAttributeBuilder.js (hunk 2 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/buffers/elementConstantAttributeBuilder.js]: Done.

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js (hunk 3 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]: Done.

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js (hunk 4 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]: Removed.

File: packages/d3fc-webgl/src/buffers/baseAttributeBuilder.js (hunk 1 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/buffers/baseAttributeBuilder.js]: I've changed to `hasPropertyChanged` (thus inverting the boolean).

File: packages/d3fc-webgl/src/series/glLine.js [hunk 6]
```diff
@@ -59,12 +70,26 @@ export default () => {
         }
     );
 
-    const program = programBuilder()
-        .mode(drawModes.TRIANGLES)
-        .verticesPerElement(12);
+    const elementIndices = elementIndicesBuilder().data([
```
Comment (chrisprice) [inline packages/d3fc-webgl/src/series/glLine.js:66]: I know it's not part of this change technically but could we make use of `elementIndices` to remove the duplication in corner?

File: packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js (hunk 5 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/buffers/vertexConstantAttributeBuilder.js]: Sure, will do.

File: packages/d3fc-webgl/src/series/glLine.js (hunk 6 above)
Comment (jamesmporter) [reply packages/d3fc-webgl/src/series/glLine.js:66]: Yes, I'll put that in.

**Review by chrisprice — CHANGES_REQUESTED** [review]
As discussed please can you re-introduce `drawArrays`

**Comment by jamesmporter** [comment]
@chrisprice I've removed the hack in the previous push, making this a lot cleaner.

**Review by chrisprice — APPROVED** [review]

**Review by OliverForeman — APPROVED** [review]

**Pull request merged**